pkg/controller/v1beta1/inferenceservice/reconcilers/keda/keda_reconciler.go

```diff
@@ -0,0 +1,212 @@
+limitations under the License.
+*/
+
+package keda
+
+import (
+	"context"
+	"strconv"
+
+	kedav1alpha1 "github.com/kedacore/keda/v2/apis/keda/v1alpha1"
+	"github.com/kserve/kserve/pkg/apis/serving/v1beta1"
+	"github.com/kserve/kserve/pkg/constants"
+	"github.com/pkg/errors"
+	"google.golang.org/protobuf/proto"
+	autoscalingv2 "k8s.io/api/autoscaling/v2"
+	corev1 "k8s.io/api/core/v1"
+	apierr "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/util/retry"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+	logf "sigs.k8s.io/controller-runtime/pkg/log"
+)
+
+var log = logf.Log.WithName("KedaReconciler")
+
+type KedaReconciler struct {
+	client       client.Client
+	scheme       *runtime.Scheme
+	ScaledObject *kedav1alpha1.ScaledObject
+	componentExt *v1beta1.ComponentExtensionSpec
+}
+
+func NewKedaReconciler(client client.Client,
+	clientset kubernetes.Interface,
+	scheme *runtime.Scheme,
+	componentMeta metav1.ObjectMeta,
+	componentExt *v1beta1.ComponentExtensionSpec) *KedaReconciler {
+	return &KedaReconciler{
+		client:       client,
+		scheme:       scheme,
+		ScaledObject: createKedaScaledObject(clientset, componentMeta, componentExt),
+		componentExt: componentExt,
+	}
+}
+
+func getKedaMetrics(metricConfig *v1beta1.MetricsConfig, metadata metav1.ObjectMeta,
+	componentExt *v1beta1.ComponentExtensionSpec) []kedav1alpha1.ScaleTriggers {
+	var triggers []kedav1alpha1.ScaleTriggers
+	var serverAddress string
+
+	// Default values
+	triggerType := string(corev1.ResourceCPU)
+	metricType := autoscalingv2.UtilizationMetricType
+	scaleTarget := int(constants.DefaultCPUUtilization)
+
+	// Get metric configuration from metricConfig (configmap)
+	if metricConfig != nil && metricConfig.MetricBackend != "" {
+		triggerType = metricConfig.MetricBackend
+		serverAddress = metricConfig.ServerAddress
+	}
+	// override metric configuration from componentExtension if it is set
+	if componentExt.ScaleMetric != nil {
+		triggerType = string(*componentExt.ScaleMetric)
+	}
+	if componentExt.ScaleMetricType != nil {
+		metricType = *componentExt.ScaleMetricType
+	}
+	if componentExt.ScaleTarget != nil {
+		scaleTarget = *componentExt.ScaleTarget
+	}
+	// override metric configuration from componentExtension.ScalerSpec if it is set
+	if componentExt.ScalerSpec != nil {
+		if componentExt.ScalerSpec.ScaleMetric != nil {
+			triggerType = string(*componentExt.ScalerSpec.ScaleMetric)
+		}
+		if componentExt.ScalerSpec.ServerAddress != "" {
+			serverAddress = componentExt.ScalerSpec.ServerAddress
+		}
+		if componentExt.ScalerSpec.ScaleMetricType != nil {
+			scaleTarget = *componentExt.ScalerSpec.ScaleTarget
+		}
+		if componentExt.ScalerSpec.ScaleMetricType != nil {
+			metricType = *componentExt.ScalerSpec.ScaleMetricType
```

Comment:
```
	// Type of metric to use. Options are Utilization, or AverageValue.
	// +optional
	ScaleMetricType *v2.MetricTargetType `json:"scaleMetricType,omitempty"`
```
Afaik you cannot have CPU with `AverageValueMetricType` should we validate early and fail fast?